### ISSUE TITLE
hostconfig.json and config.v2.json high write i/o

### ISSUE THREAD

**Issue by tambetliiv** [moby/moby#40183]
**Description**

For some reason docker daemon is not able to kill containers and starts to make high write io to `hostconfig.json` and `config.v2.json` files. 

**Steps to reproduce the issue:**
1.  Don't know what causes this issue so can't reproduce but it appears more and more often recently.

**Describe the results you received:**

Node starts to alert high load and/or high memory usage and when checking there is high io:
```
# iostat 3
Linux 4.19.0-6-amd64 (k8s-node)     11/06/2019      _x86_64_        (7 CPU)

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
          16.39    0.00    3.61    0.45    0.04   79.50

Device             tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
vda              74.44      1196.74       350.94  190335884   55815612
dm-0              7.34       360.95         1.22   57407100     193465
dm-1             53.67       830.98       373.68  132163558   59431822
dm-2              5.64         4.79        48.14     761615    7655817

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           5.87    0.00    5.63    8.30    0.05   80.16

Device             tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
vda            1702.00         2.67      5284.50          8      15853
dm-0              0.33         2.67         0.00          8          0
dm-1           1218.67         0.00      7833.83          0      23501
dm-2              6.67         0.00        64.00          0        192

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           4.31    0.00    4.50    9.20    0.05   81.94

Device             tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
vda            1694.33         0.00      5204.50          0      15613
dm-0              0.00         0.00         0.00          0          0
dm-1           1216.00         0.00      7782.67          0      23348
dm-2              0.00         0.00         0.00          0          0

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           4.86    0.00    5.05    8.86    0.05   81.19

Device             tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
vda            1604.00         9.33      4925.50         28      14776
dm-0              0.00         0.00         0.00          0          0
dm-1           1152.33         9.33      7381.50         28      22144
dm-2              0.00         0.00         0.00          0          0
```

Then drilling down using `fatrace` we find out that `hostconfig.json` and `config.v2.json` are written to: 
```
# fatrace -t --filter=W -s 60 -o fatrace-write.log
# cut -d' ' -f 4 fatrace-write.log | sort -nr -k 1 | uniq -c | sort -nr -k 1 | head
   1406 /var/lib/docker/containers/4ef33d6ecfcd27eda6e36f9726d0c66222309cd653007e0ef8f1a34dd4b78066/hostconfig.json
   1375 /var/lib/docker/containers/72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699/hostconfig.json
   1343 /var/lib/docker/containers/72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699/config.v2.json
   1341 /var/lib/docker/containers/4ef33d6ecfcd27eda6e36f9726d0c66222309cd653007e0ef8f1a34dd4b78066/config.v2.json
   1259 /var/lib/docker/containers/20077ce975b4f8494360f65d2e55ac2dfd3253efbdd43aa47cfd8a56efd0a6e4/hostconfig.json
   1242 /var/lib/docker/containers/20077ce975b4f8494360f65d2e55ac2dfd3253efbdd43aa47cfd8a56efd0a6e4/config.v2.json
   1138 /var/lib/docker/containers/9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384/hostconfig.json
   1075 /var/lib/docker/containers/9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384/config.v2.json
```

Systemd log also shows that docker is trying to kill containers but unable to do so:
```
# systemctl status docker -l --no-pager
● docker.service - Docker Application Container Engine
   Loaded: loaded (/etc/systemd/system/docker.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─docker-dns.conf, docker-options.conf
   Active: active (running) since Mon 2019-11-04 14:19:47 EET; 1 day 20h ago
     Docs: http://docs.docker.com
 Main PID: 577 (dockerd)
    Tasks: 0
   Memory: 52.2M
   CGroup: /system.slice/docker.service
           └─577 /usr/bin/dockerd --data-root=/var/lib/docker --log-opt max-size=50m --log-opt max-file=5 --iptables=true --data-root=/var/lib/docker --log-opt max-size=50m --log-opt max-file=5 --dns 10.233.0.3 --dns 192.168.1.1 --dns-search default.svc.example.com --dns-search svc.example.com --dns-opt ndots:2 --dns-opt timeout:2 --dns-opt attempts:2

Nov 06 10:28:33 k8s-node dockerd[577]: time="2019-11-06T10:28:33.578517689+02:00" level=info msg="Container 9fc9f4c1e256 failed to exit within 10 seconds of kill - trying direct SIGKILL"
Nov 06 10:28:43 k8s-node dockerd[577]: time="2019-11-06T10:28:43.684813540+02:00" level=info msg="Container 72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699 failed to exit within 10 seconds of signal 15 - using the force"
Nov 06 10:28:53 k8s-node dockerd[577]: time="2019-11-06T10:28:53.759684308+02:00" level=info msg="Container 72d5e7108eb3 failed to exit within 10 seconds of kill - trying direct SIGKILL"
Nov 06 10:30:09 k8s-node dockerd[577]: time="2019-11-06T10:30:09.537089218+02:00" level=info msg="Container 20077ce975b4f8494360f65d2e55ac2dfd3253efbdd43aa47cfd8a56efd0a6e4 failed to exit within 30 seconds of signal 15 - using the force"
Nov 06 10:30:19 k8s-node dockerd[577]: time="2019-11-06T10:30:19.654768778+02:00" level=info msg="Container 20077ce975b4 failed to exit within 10 seconds of kill - trying direct SIGKILL"
Nov 06 10:30:29 k8s-node dockerd[577]: time="2019-11-06T10:30:29.787106806+02:00" level=info msg="Container 4ef33d6ecfcd27eda6e36f9726d0c66222309cd653007e0ef8f1a34dd4b78066 failed to exit within 10 seconds of signal 15 - using the force"
Nov 06 10:30:39 k8s-node dockerd[577]: time="2019-11-06T10:30:39.876904454+02:00" level=info msg="Container 4ef33d6ecfcd failed to exit within 10 seconds of kill - trying direct SIGKILL"
Nov 06 10:30:43 k8s-node dockerd[577]: time="2019-11-06T10:30:43.493346555+02:00" level=info msg="Container 9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384 failed to exit within 30 seconds of signal 15 - using the force"
Nov 06 10:30:53 k8s-node dockerd[577]: time="2019-11-06T10:30:53.593008482+02:00" level=info msg="Container 9fc9f4c1e256 failed to exit within 10 seconds of kill - trying direct SIGKILL"
Nov 06 10:31:03 k8s-node dockerd[577]: time="2019-11-06T10:31:03.719245791+02:00" level=info msg="Container 72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699 failed to exit within 10 seconds of signal 15 - using the force"
```

Issue resolves after docker service restart or host reboot.

**Describe the results you expected:**

Docker daemon killing containers successfully and not creating high io.

**Additional information you deem important (e.g. issue happens only occasionally):**

It happens occasinally. Have not found any correlation about containers that are causing this issue.

**Output of `docker inspect`:**

```
# docker inspect 9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384
[
    {
        "Id": "9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384",
        "Created": "2019-11-06T06:23:41.859482225Z",
        "Path": "docker-entrypoint.sh",
        "Args": [
            "node",
            "build/main.js"
        ], 
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 25545,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2019-11-06T06:23:42.19151414Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:219e65eab917b499b4ad7670a262a2d56d647c03e00e5ce1ee82649e47ae4d7d",
        "ResolvConfPath": "/var/lib/docker/containers/72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699/hostname",
        "HostsPath": "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/etc-hosts",
        "LogPath": "/var/lib/docker/containers/9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384/9fc9f4c1e25688cd7445c8ca1aca4d3232bc0b8a34705be9eaf723bd706a9384-json.log",
        "Name": "<omitted>",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "docker-default",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/volumes/kubernetes.io~secret/default-token-j6rbq:/var/run/secrets/kubernetes.io/serviceaccount:ro",
                "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/etc-hosts:/etc/hosts",
                "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/containers/<omitted>/20a4d31a:/dev/termination-log"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {
                    "max-file": "5",
                    "max-size": "50m"
                }
            },
            "NetworkMode": "container:72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699",
            "PortBindings": null,
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "container:72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 997,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": [
                "seccomp=unconfined"
            ],
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 5,
            "Memory": 314572800,
            "NanoCpus": 0,
            "CgroupParent": "/kubepods/burstable/pod8355e83b-d877-4d23-8a35-5f4dbb91f05b",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 100000,
            "CpuQuota": 35000,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "DeviceCgroupRules": null,
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 314572800,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/asound",
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/a63e4c591524c5288fdb1739c7f5e66c142b0aaed379b88880ef4c0dd042ee92-init/diff:/var/lib/docker/overlay2/f539c4009527df1b85276dd1247869473c098078dcd362be7b700f0a4ccbd488/diff:/var/lib/docker/overlay2/6b4d16c0a8930b5378cfc500fda6d19a98fad15c5e1a7f2be6941b2ba68b6e7f/diff:/var/lib/docker/overlay2/85cfb62c3cf2487e4424c4c2fee74f4585c8d37c6d1a367fda23d381a18d08b7/diff:/var/lib/docker/overlay2/cc2d39c1f1ef7ce2d0bc4bb0aeacc34fa857aa756c2ea1788d2179e2e6023562/diff:/var/lib/docker/overlay2/561b4149b33a3cdff561a48a648f233cd683309fc968a97831ad6b6dc71d8ac4/diff:/var/lib/docker/overlay2/97cbcc038f3122db5a60c40652e5d27ab4c328a6aba1ba83d0362bd1b60dadd2/diff:/var/lib/docker/overlay2/a149b5b02930387159120be673caf881e38fc787cc34db912c15f75d929108b1/diff",
                "MergedDir": "/var/lib/docker/overlay2/a63e4c591524c5288fdb1739c7f5e66c142b0aaed379b88880ef4c0dd042ee92/merged",
                "UpperDir": "/var/lib/docker/overlay2/a63e4c591524c5288fdb1739c7f5e66c142b0aaed379b88880ef4c0dd042ee92/diff",
                "WorkDir": "/var/lib/docker/overlay2/a63e4c591524c5288fdb1739c7f5e66c142b0aaed379b88880ef4c0dd042ee92/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/volumes/kubernetes.io~secret/default-token-j6rbq",
                "Destination": "/var/run/secrets/kubernetes.io/serviceaccount",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/etc-hosts",
                "Destination": "/etc/hosts",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/var/lib/kubelet/pods/8355e83b-d877-4d23-8a35-5f4dbb91f05b/containers/<omitted>/20a4d31a",
                "Destination": "/dev/termination-log",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "<omitted>",
            "Domainname": "",
            "User": "node",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "3000/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                <omitted>
            ],
            "Cmd": [
                "node",
                "build/main.js"
            ],
            "Healthcheck": {
                "Test": [
                    "NONE"
                ]
            },
            "Image": "sha256:219e65eab917b499b4ad7670a262a2d56d647c03e00e5ce1ee82649e47ae4d7d",
            "Volumes": null,
            "WorkingDir": "/app",
            "Entrypoint": [
                "docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {
                "annotation.io.kubernetes.container.hash": "3cd8436c",
                "annotation.io.kubernetes.container.ports": "[{\"name\":\"ppublic\",\"containerPort\":3000,\"protocol\":\"TCP\"}]",
                "annotation.io.kubernetes.container.restartCount": "4",
                "annotation.io.kubernetes.container.terminationMessagePath": "/dev/termination-log",
                "annotation.io.kubernetes.container.terminationMessagePolicy": "File",
                "annotation.io.kubernetes.pod.terminationGracePeriod": "30",
                "io.kubernetes.container.logpath": "/var/log/pods/<omitted>/<omitted>/4.log",
                "io.kubernetes.container.name": "<omitted>",
                "io.kubernetes.docker.type": "container",
                "io.kubernetes.pod.name": "<omitted>",
                "io.kubernetes.pod.namespace": "<omitted>",
                "io.kubernetes.pod.uid": "8355e83b-d877-4d23-8a35-5f4dbb91f05b",
                "io.kubernetes.sandbox.id": "72d5e7108eb3751ad4c7e05711cbf53a033a2771b63df945055565343a602699"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {}
        }
    }
]

```

**Output of `docker version`:**

```
# docker version
Client: Docker Engine - Community
 Version:           19.03.2
 API version:       1.39 (downgraded from 1.40)
 Go version:        go1.12.8
 Git commit:        6a30dfc
 Built:             Thu Aug 29 05:29:29 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.9
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.13
  Git commit:       039a7df
  Built:            Wed Sep  4 16:19:38 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
# docker info
Client:
 Debug Mode: false

Server:
 Containers: 27
  Running: 15
  Paused: 0
  Stopped: 12
 Images: 56
 Server Version: 18.09.9
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-6-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 7
 Total Memory: 15.67GiB
 Name: kubernetes-node
 ID: 5QRU:B74U:7NJ6:5ZBB:GE76:SXYV:6SEI:3D4F:CYCX:JTMV:3JIL:C23N
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

All docker hosts that have this issue are Kubernetes worker nodes. Hosts themselves are KVM debian VMs.

**Comment by tambetliiv** [comment]
Still having this problem with the latest version also ```node5 ~# docker version
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea838
 Built:             Wed Nov 13 07:25:38 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea838
  Built:            Wed Nov 13 07:24:09 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
````

Seems similar/identical to https://github.com/docker/for-linux/issues/670 because when checking `docker events` there were a lot of `container oom` messages spamming.

Also `docker rm --force` helps which was also suggested [here](https://github.com/docker/for-linux/issues/670#issuecomment-501943117)

**Comment by tambetliiv** [comment]
https://github.com/moby/moby/issues/38352 suggested that the bug maybe in kernel.

Then found https://github.com/torvalds/linux/commit/f9c645621a28e37813a1de96d9cbd89cde94a1e4 which was included in 4.19 kernel from [4.19.77](https://cdn.kernel.org/pub/linux/kernel/v4.x/ChangeLog-4.19.77) (and in newer ones since [5.2.19](https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.2.19) and [5.3.4](http://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.3.4))

Until now i had only tested this issue with kernels older than 4.19.77.

To verify this I compiled kernels 4.19.76 and 4.19.77 on my debian buster host (which is using XFS filesystem) to verify that [this](https://github.com/torvalds/linux/commit/f9c645621a28e37813a1de96d9cbd89cde94a1e4) fix really helps. And it did, with kernel 4.19.76 i was able to reproduce this oom issue after reboot with no problems wih command 
```
docker run --name mem_test --init -d --rm -m 50m --memory-swap 50m python python -c "import itertools; list(itertools.count())"
```
and after booting up kernel 4.19.77 wasn't able to reproduce this issue anymore.

So to fix this issue upgrade your kernel to 4.19.77, 5.2.19 and 5.3.4 or above.

**Comment by thaJeztah** [comment]
Thanks for the update!
